Add code to provide proper IOB site pin names for bonded IOB's.  Only

ttown523 opened the pull request:
Add code to provide proper IOB site pin names for bonded IOB's.  Only affects the .txt file.

This is to enable conversion from Vivado to RS2 - the placement.xdc and placement.xdc.txt files have the wrong site names for bonded IOB's.

ttown523 committed:
Add code to provide proper IOB site pin names for bonded IOB's.  Only
affects the .txt file.

ttown523 committed:
Merge remote-tracking branch 'origin/master'

ttown523 commented:
Why does this pull request show the result of 2 commits?  Only the top one was intended.  The bottom commit looks significant and doesn't seem to be what we want.

bradselw commented:
Looks like it includes the merge of my master to yours. Doesn't affect anything on my end.

bradselw commented on tincr/io/design/tincr_checkpoints.tcl:
```diff
@@ -127,10 +128,23 @@ proc ::tincr::write_placement_xdc {args} {
             
             set group [get_property PRIMITIVE_GROUP $cell]
-            if {$group == "LUT" || $group == "INV" || $group == "BUF"} {
+            if {$group == "LUT" || $group == "INV" || $group == "BUF"} {	
```
Remove whitespace at the end of the line

bradselw commented on tincr/io/design/tincr_checkpoints.tcl:
```diff
@@ -127,10 +128,23 @@ proc ::tincr::write_placement_xdc {args} {
+
+	    # To match the xdlrc file generated from TINCR's "write_xdlrc", IOB sites that are bonded
+	    # have been updated to use the package pin name rather than the actual site name.  
+	    # The Vivado .xdc files have not been changed, only the txt files used for parsing a design into rapidSmith2
```
I think that the logic for writing the RS2 txt files should be split out into a separate function for a few reasons:
1. It would reduce the amount of work that the function needs to do to generate the desired data
2. When you run write_tcp, 99% of the time you only want one of the two files - if they were split, it would be one file per function
3. Easier to maintain

bradselw commented on tincr/io/design/tincr_checkpoints.tcl:
```diff
@@ -127,10 +128,23 @@ proc ::tincr::write_placement_xdc {args} {
+	    # The Vivado .xdc files have not been changed, only the txt files used for parsing a design into rapidSmith2
+            set sitename [get_property LOC $cell]
+
+            if { [regexp {.*IOB*} $sitename]} {
```
I don't think it is a good idea to key off of a substring ("IOB") in the name like this. There's no contract between a site's name and its type. Better to use the site's properties instead. Consider how write_xdlrc does it:

```
    foreach site $sites {
        set state "internal"
        set name [get_property NAME $site]
        if {[get_property IS_PAD $site]} {
            if {[get_property IS_BONDED $site]} {
                set state "bonded"
                set name [get_property NAME [get_package_pins -quiet -of_object $site]]
            } else {
                set state "unbonded"
            }
        }
```

nelsobe replied:
I agree with Brad’s suggestion.  It does not rely on names to differentiate.

From: Brad White <notifications@github.com<mailto:notifications@github.com>>
Reply-To: bradselw/tincr <reply@reply.github.com<mailto:reply@reply.github.com>>
Date: Tuesday, February 9, 2016 at 1:20 AM
To: bradselw/tincr <tincr@noreply.github.com<mailto:tincr@noreply.github.com>>
Subject: Re: [tincr] Add code to provide proper IOB site pin names for bonded IOB's. Only (#15)

In tincr/io/design/tincr_checkpoints.tclhttps://github.com/bradselw/tincr/pull/15#discussion_r52277756:

> @@ -127,10 +128,23 @@ proc ::tincr::write_placement_xdc {args} {
>              puts $xdc "set_property BEL [get_property BEL $cell] [get_cells {[get_name $cell]}]"
>              puts $xdc "set_property LOC [get_property LOC $cell] [get_cells {[get_name $cell]}]"
>              puts $txt "BEL [get_property BEL $cell] [get_cells [get_name $cell]]"
> -            puts $txt "LOC [get_name $cell] [get_property LOC $cell] [get_property BEL $cell] [get_tile -of [get_sites -of $cell]]"
>   +
> -         # To match the xdlrc file generated from TINCR's "write_xdlrc", IOB sites that are bonded
> -         # have been updated to use the package pin name rather than the actual site name.
> -         # The Vivado .xdc files have not been changed, only the txt files used for parsing a design into rapidSmith2
> -            set sitename [get_property LOC $cell]
>   +
> -            if { [regexp {._IOB_} $sitename]} {

I don't think it is a good idea to key off of a substring ("IOB") in the name like this. There's no contract between a site's name and its type. Better to use the site's properties instead. Consider how write_xdlrc does it:

```
foreach site $sites {
    set state "internal"
    set name [get_property NAME $site]
    if {[get_property IS_PAD $site]} {
        if {[get_property IS_BONDED $site]} {
            set state "bonded"
            set name [get_property NAME [get_package_pins -quiet -of_object $site]]
        } else {
            set state "unbonded"
        }
    }
```

—
Reply to this email directly or view it on GitHubhttps://github.com/bradselw/tincr/pull/15/files#r52277756.

---

---

ttown523 committed:
Seperated TINCR checkpoints from RapidSmith2 checkpoints. This will
allow us to create RapidSmith2 checkpoints without modifiying the old
TINCR checkpoint code.

ttown523 commented:
Hi Brad. I have taken your suggestions, and submitted a new pull request. Take a look and let me know what you think. Thanks!
